src/box/memtx_tx.c

```diff
@@ -2996,13 +2961,87 @@ memtx_tx_delete_gap(struct gap_item_base *item)
 	memtx_tx_mempool_free(item->txn, pool, item);
 }
 
+/**
+ * Abort all concurrent readers of the old schema of the space.
+ */
 void
-memtx_tx_invalidate_space(struct space *space, struct txn *active_txn)
+memtx_tx_abort_space_schema_readers(struct space *space, struct txn *ddl_owner)
 {
+	/* Abort all story and gap-with-sucessor readers. */
+	struct memtx_story *story;
+	rlist_foreach_entry(story, &space->memtx_stories, in_space_stories) {
+		struct tx_read_tracker *tracker, *tmp;
+		rlist_foreach_entry_safe(tracker, &story->reader_list,
+					 in_reader_list, tmp) {
+			if (tracker->reader != ddl_owner)
+				memtx_tx_abort_with_conflict(tracker->reader);
```

Comment:
We could use more fine-grained approach, but it would be more fragile. Imagine altering a primary index. Despite the fact your non-unique secondary index wasn't touched by user, it is actually rebuilt since it uses primary key as a tie-breaker. Are there more pitfalls like this? I believe we would miss one or another.

So let's simply abort all space readers. DDLs are rare actually, so it doesn't seem to be a problem.